### PR TITLE
expression: constant propagation for outer join don't prop control functions

### LINKED ISSUE #15782: planner: invalid join results when push conditions down to subquery with left outer join
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
```
create table t1 (id int);
create table t2 (id int, v varchar(10));
insert into t1 values (1);
insert into t2 values (1, 'x');
select * from (
  select ifnull(t2.v, t1.id) as t2v from t1
  left join t2 on t1.id = t2.id
) t where t2v != 'x';
```

### 2. What did you expect to see?
```
mysql> create table t1 (id int);
Query OK, 0 rows affected (0.05 sec)

mysql> create table t2 (id int, v varchar(10));
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 values (1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t2 values (1, 'x');
Query OK, 1 row affected (0.00 sec)

mysql> select * from (
    ->   select ifnull(t2.v, t1.id) as t2v from t1
    ->   left join t2 on t1.id = t2.id
    -> ) t where t2v != 'x';
Empty set (0.01 sec)
```

### 3. What did you see instead?
```
mysql> create table t1 (id int);
Query OK, 0 rows affected (0.01 sec)

mysql> create table t2 (id int, v varchar(10));
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 values (1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t2 values (1, 'x');
Query OK, 1 row affected (0.00 sec)

mysql> select * from (
    ->   select ifnull(t2.v, t1.id) as t2v from t1
    ->   left join t2 on t1.id = t2.id
    -> ) t where t2v != 'x';
+------+
| t2v  |
+------+
| 1    |
+------+
1 row in set (0.00 sec)
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
The latest master branch.

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #15782  

Problem Summary:

Wrong propagation happened.

### What is changed and how it works?

What's Changed:

Don't propagate for the control function.
We do need to check the `isNullRejected`, but this should be done after we maintain the nullable information more strictly.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Review by qw4990 — COMMENTED** [review]
LGTM.

File: expression/testdata/expression_suite_out.json [hunk 1]
```diff
@@ -274,6 +274,18 @@
           "    └─TableReader_11(Probe) 10000.00 root  data:TableFullScan_10",
           "      └─TableFullScan_10 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
         ]
+      },
+      {
+        "SQL": "explain select * from t1 left join t2 on t1.a = t2.a where ifnull(t2.b, t1.a) = 1",
+        "Result": [
+          "Selection_7 9990.00 root  eq(ifnull(test.t2.b, test.t1.a), 1)",
```
Comment (lzmhhh123) [inline expression/testdata/expression_suite_out.json:281]: [optional] Can we do twice selections? The first is pushed down, the second as the root for this one?

**Comment by SunRunAway** [comment]
Should be cherry-picked before v3.0.13

**Review by lzmhhh123 — APPROVED** [review]
LGTM

File: expression/testdata/expression_suite_out.json [hunk 2]
```diff
@@ -274,6 +274,18 @@
           "    └─TableReader_11(Probe) 10000.00 root  data:TableFullScan_10",
           "      └─TableFullScan_10 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
         ]
+      },
+      {
+        "SQL": "explain select * from t1 left join t2 on t1.a = t2.a where ifnull(t2.b, t1.a) = 1",
```
Comment (zz-jason) [inline expression/testdata/expression_suite_out.json:279]:
```sql
TiDB(root@127.0.0.1:test) > explain select * from t1 left join t2 on t1.a = t2.a where t2.b + t1.a = 1;
+------------------------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------+
| id                           | estRows  | task      | access object | operator info                                                                              |
+------------------------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------+
| HashJoin_7                   | 9990.00  | root      |               | inner join, equal:[eq(test.t1.a, test.t2.a)], other cond:eq(plus(test.t2.b, test.t1.a), 1) |
| ├─TableReader_14(Build)      | 7992.00  | root      |               | data:Selection_13                                                                          |
| │ └─Selection_13             | 7992.00  | cop[tikv] |               | eq(plus(test.t2.b, test.t2.a), 1), not(isnull(test.t2.a))                                  |
| │   └─TableFullScan_12       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                             |
| └─TableReader_11(Probe)      | 9990.00  | root      |               | data:Selection_10                                                                          |
|   └─Selection_10             | 9990.00  | cop[tikv] |               | not(isnull(test.t1.a))                                                                     |
|     └─TableFullScan_9        | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                             |
+------------------------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------+
7 rows in set (0.00 sec)
```

this query should also not be converted to `inner join`, please confirm.

**Review by zz-jason — CHANGES_REQUESTED** [review]

File: expression/testdata/expression_suite_out.json (hunk 2 above)
Comment (winoros) [reply expression/testdata/expression_suite_out.json:279]: the `t2.b+t1.a` is null rejective for t2 side. So it can be rewrited to inner join.

**Review by zz-jason — APPROVED** [review]

**Pull request merged**